File: test_drive/lib/widgets/hybrid_editor/core/block_node.dart
import 'package:flutter/foundation.dart';

/// Enum representing all possible block types in the editor
enum BlockType {
  paragraph,
  heading1,
  heading2,
  heading3,
  bulletList,
  numberedList,
  codeBlock,
  mathBlock,
}

/// Immutable node representing a block of content in the document.
/// 
/// Each block has a unique ID, type, and content. Code blocks also
/// store the programming language for syntax highlighting.
@immutable
class BlockNode {
  final String id;
  final BlockType type;
  final String content;
  final String? language; // For code blocks
  final int listLevel; // For list blocks
  final Map<String, dynamic> attributes;

  const BlockNode({
    required this.id,
    required this.type,
    required this.content,
    this.language,
    this.listLevel = 0,
    this.attributes = const {},
  });

  /// Creates a new paragraph block with generated ID
  factory BlockNode.paragraph({String content = ''}) {
    return BlockNode(
      id: generateId(),
      type: BlockType.paragraph,
      content: content,
    );
  }

  /// Creates a new heading block
  factory BlockNode.heading({
    required int level,
    String content = '',
  }) {
    assert(level >= 1 && level <= 3, 'Heading level must be 1, 2, or 3');
    final type = switch (level) {
      1 => BlockType.heading1,
      2 => BlockType.heading2,
      3 => BlockType.heading3,
      _ => BlockType.heading1,
    };
    return BlockNode(
      id: generateId(),
      type: type,
      content: content,
    );
  }

  /// Creates a new list block
  factory BlockNode.list({
    required bool numbered,
    String content = '',
    int listLevel = 0,
  }) {
    return BlockNode(
      id: generateId(),
      type: numbered ? BlockType.numberedList : BlockType.bulletList,
      content: content,
      listLevel: listLevel,
    );
  }

  /// Creates a new code block
  factory BlockNode.codeBlock({
    String content = '',
    String? language,
  }) {
    return BlockNode(
      id: generateId(),
      type: BlockType.codeBlock,
      content: content,
      language: language,
    );
  }

  factory BlockNode.mathBlock({
    String content = '',
  }) {
    return BlockNode(
      id: generateId(),
      type: BlockType.mathBlock,
      content: content,
    );
  }

  /// Creates a copy of this block with updated fields
  BlockNode copyWith({
    String? id,
    BlockType? type,
    String? content,
    String? language,
    int? listLevel,
    Map<String, dynamic>? attributes,
  }) {
    return BlockNode(
      id: id ?? this.id,
      type: type ?? this.type,
      content: content ?? this.content,
      language: language ?? this.language,
      listLevel: listLevel ?? this.listLevel,
      attributes: attributes ?? this.attributes,
    );
  }

  /// Converts the block to its markdown representation
  String toMarkdown() {
    final listIndentation = '${'  ' * listLevel}';
    switch (type) {
      case BlockType.heading1:
        return '# $content';
      case BlockType.heading2:
        return '## $content';
      case BlockType.heading3:
        return '### $content';
      case BlockType.bulletList:
        return '$listIndentation- $content';
      // TODO: make the list item number part of the node, not part of the config
      case BlockType.numberedList:
        return '${listIndentation}1. $content';
      case BlockType.codeBlock:
        final lang = language ?? '';
        return '```$lang\n$content\n```';
      case BlockType.mathBlock:
        return '\$\$\n$content\n\$\$';
      case BlockType.paragraph:
        return content;
    }
  }

  /// Gets the heading level if this is a heading block
  int? get headingLevel => switch (type) {
    BlockType.heading1 => 1,
    BlockType.heading2 => 2,
    BlockType.heading3 => 3,
    _ => null,
  };

  BlockType get blockType => type;

  /// Whether this block is a code block
  bool get isCodeBlock => type == BlockType.codeBlock;

  /// Whether this block is a list item
  bool get isListItem => 
    type == BlockType.bulletList || type == BlockType.numberedList;

  /// Whether this block is a heading
  bool get isHeading => 
    type == BlockType.heading1 || 
    type == BlockType.heading2 || 
    type == BlockType.heading3;

  static int _idCounter = 0;
  static String generateId() => 'block_${++_idCounter}_${DateTime.now().microsecondsSinceEpoch}';

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is BlockNode &&
          runtimeType == other.runtimeType &&
          id == other.id &&
          type == other.type &&
          content == other.content &&
          listLevel == other.listLevel &&
          language == other.language;

  @override
  int get hashCode => Object.hash(id, type, content, listLevel, language);

  @override
  String toString() => 'BlockNode(id: $id, type: $type, content: "${content.length > 20 ? '${content.substring(0, 20)}...' : content}", listLevel: $listLevel, language: $language)';
}

File: native/components/editor/core/BlockNode.ts
/// Enum representing all possible block types in the editor
export enum BlockType {
  paragraph = 'paragraph',
  heading1 = 'heading1',
  heading2 = 'heading2',
  heading3 = 'heading3',
  bulletList = 'bulletList',
  numberedList = 'numberedList',
  codeBlock = 'codeBlock',
  mathBlock = 'mathBlock',
  image = 'image',
}

/// Immutable node representing a block of content in the document.
///
/// Each block has a unique ID, type, and content. Code blocks also
/// store the programming language for syntax highlighting.
export interface BlockNode {
  readonly id: string;
  readonly type: BlockType;
  readonly content: string;
  readonly language?: string; // For code blocks
  readonly listLevel: number; // For list blocks
  readonly attributes: Record<string, unknown>;
}

let idCounter = 0;

function generateId(): string {
  return `block_${++idCounter}_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
}

/// Creates a new paragraph block with generated ID
export function createParagraphBlock(content = ''): BlockNode {
  return {
    id: generateId(),
    type: BlockType.paragraph,
    content,
    listLevel: 0,
    attributes: {},
  };
}

/// Creates a new heading block
export function createHeadingBlock(level: BlockType.heading1 | BlockType.heading2 | BlockType.heading3, content = ''): BlockNode {
  return {
    id: generateId(),
    type: level,
    content,
    listLevel: 0,
    attributes: {},
  };
}

/// Creates a new list block
export function createListBlock(
  numbered: boolean,
  content = '',
  listLevel = 0,
): BlockNode {
  return {
    id: generateId(),
    type: numbered ? BlockType.numberedList : BlockType.bulletList,
    content,
    listLevel,
    attributes: {},
  };
}

/// Creates a new code block
export function createCodeBlock(content = '', language?: string): BlockNode {
  return {
    id: generateId(),
    type: BlockType.codeBlock,
    content,
    language,
    listLevel: 0,
    attributes: {},
  };
}

/// Creates a new math block
export function createMathBlock(content = ''): BlockNode {
  return {
    id: generateId(),
    type: BlockType.mathBlock,
    content,
    listLevel: 0,
    attributes: {},
  };
}

/// Creates a new image block
export function createImageBlock(path: string): BlockNode {
  return {
    id: generateId(),
    type: BlockType.image,
    content: path,
    listLevel: 0,
    attributes: {},
  };
}

/// Creates a copy of this block with updated fields
export function copyBlock(
  block: BlockNode,
  updates: Partial<BlockNode>,
): BlockNode {
  return {
    id: updates.id ?? block.id,
    type: updates.type ?? block.type,
    content: updates.content ?? block.content,
    language: updates.language ?? block.language,
    listLevel: updates.listLevel ?? block.listLevel,
    attributes: updates.attributes ?? block.attributes,
  };
}

/// Converts the block to its markdown representation
export function blockToMarkdown(block: BlockNode, listNumber?: number): string {
  const listIndentation = '  '.repeat(block.listLevel);

  switch (block.type) {
    case BlockType.heading1:
      return `# ${block.content}`;
    case BlockType.heading2:
      return `## ${block.content}`;
    case BlockType.heading3:
      return `### ${block.content}`;
    case BlockType.bulletList:
      return `${listIndentation}- ${block.content}`;
    case BlockType.numberedList:
      return `${listIndentation}${listNumber ?? 1}. ${block.content}`;
    case BlockType.codeBlock:
      const lang = block.language ?? '';
      return `\`\`\`${lang}\n${block.content}\n\`\`\``;
    case BlockType.mathBlock:
      return `$$\n${block.content}\n$$`;
    case BlockType.image:
      return `![](${block.content})`;
    case BlockType.paragraph:
      return block.content;
  }
}

/// Whether this block is a code block
export function isCodeBlock(block: BlockNode): boolean {
  return block.type === BlockType.codeBlock;
}

/// Whether this block is a list item
export function isListItem(block: BlockNode): boolean {
  return block.type === BlockType.bulletList || block.type === BlockType.numberedList;
}

/// Whether this block is a heading
export function isHeading(block: BlockNode): boolean {
  return (
    block.type === BlockType.heading1 ||
    block.type === BlockType.heading2 ||
    block.type === BlockType.heading3
  );
}

